File: sdks/java/src/main/java/co/doow/track/management/Management.java
package co.doow.track.management;

import co.doow.track.DoowError;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;

public class Management {
    private final String apiKey;
    private final ManagementOptions options;
    private final ObjectMapper mapper;

    private final AppsResource apps;
    private final ContractsResource contracts;
    private final LicensesResource licenses;
    private final MetricsResource metrics;
    private final ExpensesResource expenses;

    public Management(String apiKey) {
        this(apiKey, new ManagementOptions());
    }

    public Management(String apiKey, ManagementOptions options) {
        String envKey = System.getenv("DOOW_TRACK_API_KEY");
        this.apiKey = envKey != null ? envKey : apiKey;
        this.options = options;

        if (this.apiKey == null || !this.apiKey.startsWith("dk_")) {
            throw new DoowError("Invalid API key format. Must start with 'dk_'.");
        }

        this.mapper = new ObjectMapper();
        this.mapper.registerModule(new JavaTimeModule());
        this.mapper.setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);
        this.mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

        this.apps = new AppsResource(this);
        this.contracts = new ContractsResource(this);
        this.licenses = new LicensesResource(this);
        this.metrics = new MetricsResource(this);
        this.expenses = new ExpensesResource(this);
    }

    public AppsResource apps() { return apps; }
    public ContractsResource contracts() { return contracts; }
    public LicensesResource licenses() { return licenses; }
    public MetricsResource metrics() { return metrics; }
    public ExpensesResource expenses() { return expenses; }

    <T> T request(String method, String path, Object body, com.fasterxml.jackson.databind.JavaType responseType) {
        return requestInternal(method, path, body, responseType);
    }

    <T> T request(String method, String path, Object body, Class<T> responseType) {
        return requestInternal(method, path, body, mapper.constructType(responseType));
    }

    @SuppressWarnings("unchecked")
    private <T> T requestInternal(String method, String path, Object body, com.fasterxml.jackson.databind.JavaType responseType) {
        String url = options.getEndpoint().replaceAll("/$", "") + "/sdk" + path;

        if (options.isDebug()) {
            System.err.println("[doow/management] " + method + " " + path);
        }

        for (int attempt = 0; attempt <= options.getRetryCount(); attempt++) {
            try {
                HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
                conn.setRequestMethod(method);
                conn.setConnectTimeout(options.getTimeoutMs());
                conn.setReadTimeout(options.getTimeoutMs());
                conn.setRequestProperty("Authorization", "Bearer " + apiKey);
                conn.setRequestProperty("Content-Type", "application/json");

                if (body != null && !method.equals("GET") && !method.equals("DELETE")) {
                    conn.setDoOutput(true);
                    String json = mapper.writeValueAsString(body);
                    conn.getOutputStream().write(json.getBytes("UTF-8"));
                }

                int status = conn.getResponseCode();

                if (status >= 200 && status < 300) {
                    if (responseType == null || responseType.getRawClass() == Void.class) {
                        return null;
                    }
                    InputStream is = conn.getInputStream();
                    return mapper.readValue(is, responseType);
                }

                if (status >= 500 && attempt < options.getRetryCount()) {
                    Thread.sleep((long) Math.pow(2, attempt) * 1000);
                    continue;
                }

                String errorBody = readStream(conn.getErrorStream());
                throw new DoowError("API error: " + errorBody, status);

            } catch (IOException e) {
                if (attempt < options.getRetryCount()) {
                    try {
                        Thread.sleep((long) Math.pow(2, attempt) * 1000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new DoowError("Request interrupted", 0);
                    }
                    continue;
                }
                throw new DoowError("Request failed: " + e.getMessage(), 0);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new DoowError("Request interrupted", 0);
            }
        }

        throw new DoowError("Max retries exceeded");
    }

    void delete(String path) {
        request("DELETE", path, null, Void.class);
    }

    ObjectMapper getMapper() {
        return mapper;
    }

    private String readStream(InputStream is) throws IOException {
        if (is == null) return "";
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }
            return sb.toString();
        }
    }
}

File: sdks/kotlin/src/main/kotlin/co/doow/track/management/Management.kt
package co.doow.track.management

import co.doow.track.DoowError
import co.doow.track.types.*
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.pow

data class ManagementOptions(
    val endpoint: String = System.getenv("DOOW_TRACK_ENDPOINT") ?: "https://api.doow.co",
    val timeoutMs: Int = 30000,
    val retryCount: Int = 3,
    val debug: Boolean = System.getenv("DOOW_TRACK_DEBUG") == "true"
)

class Management(
    apiKey: String,
    private val options: ManagementOptions = ManagementOptions()
) {
    private val apiKey = System.getenv("DOOW_TRACK_API_KEY") ?: apiKey
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = false }

    val apps = AppsResource(this)
    val contracts = ContractsResource(this)
    val licenses = LicensesResource(this)
    val metrics = MetricsResource(this)
    val expenses = ExpensesResource(this)

    init {
        require(this.apiKey.startsWith("dk_")) { "Invalid API key format. Must start with 'dk_'." }
    }

    internal inline fun <reified T> request(method: String, path: String, body: Any? = null): T {
        val url = "${options.endpoint.trimEnd('/')}/sdk$path"
        log("[doow/management] $method $path")

        repeat(options.retryCount + 1) { attempt ->
            try {
                val conn = URL(url).openConnection() as HttpURLConnection
                conn.requestMethod = method
                conn.connectTimeout = options.timeoutMs
                conn.readTimeout = options.timeoutMs
                conn.setRequestProperty("Authorization", "Bearer $apiKey")
                conn.setRequestProperty("Content-Type", "application/json")

                if (body != null && method != "GET" && method != "DELETE") {
                    conn.doOutput = true
                    val jsonBody = json.encodeToString(body)
                    conn.outputStream.use { it.write(jsonBody.toByteArray()) }
                }

                val status = conn.responseCode

                if (status in 200..299) {
                    val responseBody = conn.inputStream.bufferedReader().readText()
                    return json.decodeFromString(responseBody)
                }

                if (status >= 500 && attempt < options.retryCount) {
                    Thread.sleep(2.0.pow(attempt).toLong() * 1000)
                    return@repeat
                }

                val errorBody = conn.errorStream?.bufferedReader()?.readText() ?: ""
                throw DoowError("API error: $errorBody", status)

            } catch (e: Exception) {
                if (e is DoowError) throw e
                if (attempt < options.retryCount) {
                    Thread.sleep(2.0.pow(attempt).toLong() * 1000)
                    return@repeat
                }
                throw DoowError("Request failed: ${e.message}")
            }
        }

        throw DoowError("Max retries exceeded")
    }

    internal fun delete(path: String) {
        val url = "${options.endpoint.trimEnd('/')}/sdk$path"
        log("[doow/management] DELETE $path")

        repeat(options.retryCount + 1) { attempt ->
            try {
                val conn = URL(url).openConnection() as HttpURLConnection
                conn.requestMethod = "DELETE"
                conn.connectTimeout = options.timeoutMs
                conn.readTimeout = options.timeoutMs
                conn.setRequestProperty("Authorization", "Bearer $apiKey")

                val status = conn.responseCode
                if (status in 200..299 || status == 204) return

                if (status >= 500 && attempt < options.retryCount) {
                    Thread.sleep(2.0.pow(attempt).toLong() * 1000)
                    return@repeat
                }

                val errorBody = conn.errorStream?.bufferedReader()?.readText() ?: ""
                throw DoowError("API error: $errorBody", status)
            } catch (e: Exception) {
                if (e is DoowError) throw e
                if (attempt < options.retryCount) {
                    Thread.sleep(2.0.pow(attempt).toLong() * 1000)
                    return@repeat
                }
                throw DoowError("Request failed: ${e.message}")
            }
        }
    }

    private fun log(message: String) {
        if (options.debug) System.err.println(message)
    }
}

class AppsResource(private val mgmt: Management) {
    fun list(cursor: String? = null, limit: Int = 50): PaginatedResponse<App> {
        var path = "/apps?limit=$limit"
        if (cursor != null) path += "&cursor=$cursor"
        return mgmt.request("GET", path)
    }

    fun get(id: String): App = mgmt.request("GET", "/apps/$id")

    fun create(input: CreateAppInput): App = mgmt.request("POST", "/apps", input)

    fun delete(id: String) = mgmt.delete("/apps/$id")
}

class ContractsResource(private val mgmt: Management) {
    fun listByApp(appId: String, cursor: String? = null, limit: Int = 50): PaginatedResponse<Contract> {
        var path = "/apps/$appId/contracts?limit=$limit"
        if (cursor != null) path += "&cursor=$cursor"
        return mgmt.request("GET", path)
    }

    fun get(id: String): Contract = mgmt.request("GET", "/contracts/$id")

    fun create(appId: String, input: CreateContractInput): Contract =
        mgmt.request("POST", "/apps/$appId/contracts", input)

    fun delete(id: String) = mgmt.delete("/contracts/$id")
}

class LicensesResource(private val mgmt: Management) {
    fun list(cursor: String? = null, limit: Int = 50): PaginatedResponse<License> {
        var path = "/licenses?limit=$limit"
        if (cursor != null) path += "&cursor=$cursor"
        return mgmt.request("GET", path)
    }

    fun get(id: String): License = mgmt.request("GET", "/licenses/$id")
}

class MetricsResource(private val mgmt: Management) {
    fun listByLicense(licenseId: String, cursor: String? = null, limit: Int = 50): PaginatedResponse<Metric> {
        var path = "/licenses/$licenseId/metrics?limit=$limit"
        if (cursor != null) path += "&cursor=$cursor"
        return mgmt.request("GET", path)
    }

    fun get(id: String): Metric = mgmt.request("GET", "/metrics/$id")

    fun create(licenseId: String, input: CreateMetricInput): Metric =
        mgmt.request("POST", "/licenses/$licenseId/metrics", input)

    fun delete(id: String) = mgmt.delete("/metrics/$id")
}

class ExpensesResource(private val mgmt: Management) {
    fun listByApp(appId: String, cursor: String? = null, limit: Int = 50): PaginatedResponse<Expense> {
        var path = "/apps/$appId/expenses?limit=$limit"
        if (cursor != null) path += "&cursor=$cursor"
        return mgmt.request("GET", path)
    }
}
